src/stateStream.spec.ts

```diff
@@ -0,0 +1,134 @@
+import { reducer, actionCreator } from 'rxbeach';
+import untypedTest from 'ava';
+import { marbles } from 'rxjs-marbles/ava';
+import sinon from 'sinon';
+import { stateStream, stateStreamWithErrorHandling } from 'rxbeach/stateStream';
+import { of } from 'rxjs';
+import { stubRethrowErrorGlobally } from 'rxbeach/internal/testUtils';
+
+const incrementOne = actionCreator('[increment] one');
+const decrement = actionCreator('[increment] decrement');
+const incrementMany = actionCreator<number>('[increment] many');
+
+const incrementOneHandler = sinon.spy((accumulator: number) => accumulator + 1);
+const handleOne = reducer(incrementOne, incrementOneHandler);
+const handleMany = reducer(
+  incrementMany,
+  (accumulator: number, increment) => accumulator + increment
+);
+const test = stubRethrowErrorGlobally(untypedTest);
+
+const actions = {
+  '1': incrementOne(),
+  '2': incrementMany(2),
+  d: decrement(),
+};
+const words = {
+  a: '1',
+  b: '12',
+};
+const numbers = {
+  '1': 1,
+  '2': 2,
+  '3': 3,
+  '4': 4,
+  '5': 5,
+  '6': 6,
+  '7': 7,
+};
+
+test(
+  'stateStream subscribes action$',
+  marbles(m => {
+    const defaultState = 3;
+    const action$ = m.hot('', actions);
+    const state$ = stateStream('testStream', action$, defaultState, []);
+    state$.subscribe();
+    m.expect(action$).toHaveSubscriptions(['^']);
+  })
+);
+
+test(
+  'stateStream should emit initial state synchronously upon subscription',
+  marbles(m => {
+    const defaultState = 3;
+    const action$ = m.hot('');
+    const expected$ = m.hot('3', numbers);
+    const state$ = stateStream('testStream', action$, defaultState, []);
+
+    m.expect(state$).toBeObservable(expected$);
+  })
+);
+
+test(
+  'stateStream should call reducers synchronously if they emit at once',
+  marbles(m => {
+    const action$ = m.hot('', actions);
+    const word$ = m.hot('    a', words);
+    const expected$ = m.hot('(12)', numbers);
+    const defaultState = 1;
+
+    const handleWord = reducer(
+      word$,
+      (state: number, word) => state + word.length
+    );
+    const state$ = stateStream('testStream', action$, defaultState, [
+      handleWord,
+    ]);
+
+    m.expect(state$).toBeObservable(expected$);
+  })
+);
+
+test('stateStream subscribers should only see one emission if reducers emit synchronously', async t => {
+  const defaultState = 1;
+  const action$ = of(incrementOne());
+  const state$ = stateStream('testStream', action$, defaultState, [handleOne]);
+  t.is(await state$.toPromise(), 2);
+});
+
+test('stateStream should call reducer once when there are multiple subs', async t => {
```

Comment:
```suggestion
test('stateStream should call reducer once when there are multiple subs', t => {
```
An alternative way to test this would be to do it like the test `stateStream subscribes action$`, but check that `action$` still only have one subscription when the `state$` has two.

Reply:
True, but I think this way is more explicit. I could add the test that ensures `action$` is only subscribed once, but that might be superfluous?